### PR TITLE
interchange: Allow pseudo-cells with no input pins

### PR DESCRIPTION
**Pull request by gatecat**
These are used for the LUT-as-GND-driver pseudo-pips in the Nexus arch, which will probably be required for UltraScale too.

Anyone want to give this a review? @acomodi or @tmichalak ?

### CODE REVIEW

File: fpga_interchange/pseudo_pip_model.cc [hunk 1]
```diff
@@ -378,13 +377,22 @@ void PseudoPipModel::update_site(const Context *ctx, size_t site)
...
 
-        // Map LUT input to input wire used by pseudo pip.
-        IdString input_bel_pin(bel_data.ports[bel.input_bel_pin]);
-        cell.cell_bel_pins[IdString(ctx->wire_lut->input_pins[0])].push_back(input_bel_pin);
+        if (bel.input_bel_pin == -1) {
+            // FIXME: currently assume that LUT route-throughs with no input pins are GND drivers as this is all we need
+            // for Nexus/Xilinx where Vcc is readily available and cheap This won't be true for other arches
+            cell.lut_cell.equation.resize(2);
+            cell.lut_cell.equation.set(0, false);
+            cell.lut_cell.equation.set(1, false);
```
Comment (acomodi) [inline fpga_interchange/pseudo_pip_model.cc:386]: This allows having constant GND generated by the LUT, is this correct?

File: fpga_interchange/pseudo_pip_model.cc [hunk 2]
```diff
@@ -440,7 +448,7 @@ void PseudoPipModel::update_site(const Context *ctx, size_t site)
         // that pseudo pip violates the LUT element equation.
         for (const PseudoPipBel &bel : bels) {
             const BelInfoPOD &bel_data = type_data.bel_data[bel.bel_index];
-            if (bel_data.lut_element == -1) {
+            if (bel_data.lut_element == -1 || bel.input_bel_pin == -1) {
```
Comment (acomodi) [inline fpga_interchange/pseudo_pip_model.cc:451]: I am not sure about this, but shouldn't the no-input pseudo cell be checked as well?

**Review by acomodi — COMMENTED** [review]
Comments in

File: fpga_interchange/pseudo_pip_model.cc (hunk 1 above)
Comment (gatecat) [reply fpga_interchange/pseudo_pip_model.cc:386]: as it stands yes, see the comment this is the main use case for LUT pseudo-cells with no input pins. supporting anything else will need more work on the schema side, I think

File: fpga_interchange/pseudo_pip_model.cc (hunk 2 above)
Comment (gatecat) [reply fpga_interchange/pseudo_pip_model.cc:451]: hmm, possibly, I think I need to understand how `lut_wires_unavailable` works better

Comment (gatecat) [reply fpga_interchange/pseudo_pip_model.cc]: I've now added a check that in this case, any of the bel pins is not blocked by the LUT equation, which I think is at least more correct than no check at all.

File: fpga_interchange/pseudo_pip_model.cc (hunk 1 above)
Comment (acomodi) [reply fpga_interchange/pseudo_pip_model.cc:386]: All right. As I am dealing with the LUT-thru FASM generation, I wonder how to express and more importantly identify this case. 

For instance, at the moment I am implementing the logic to identify the LUT-thrus by examining whether a LUT bel is traversed entirely in the same physical net, meaning that the LUT was indeed used as a route-thru.

In this case, however, I assume (maybe wrongly) that the net will start from the pseudoCell output, in this case, the LUT output, to reach the various sinks, therefore making the above logic unusable, as the input pins of the LUT appears as unused.

Comment (gatecat) [reply fpga_interchange/pseudo_pip_model.cc:386]: I think this special case could be identified as the net being a ground/vcc net on the LUT output, which should force the LUT function for that net to all-zeros or all-ones accordingly.

Comment (acomodi) [reply fpga_interchange/pseudo_pip_model.cc:386]: Ah, right, cause the net will be a const net, therefore yes, this case can be actually easily identified. Thanks for the clarification.

**Review by acomodi — APPROVED** [review]
LGTM

**Pull request merged**